odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+                x : `LinearSpaceVector`
+                    Element in the domain of the functional to which the
+                    gradient operator is applied. The element must have a
+                    non-zero norm
+
+                Returns
+                -------
+                `self(x)` : `LinearSpaceVector`
+                    Evaluation of the gradient operator. An element in the
+                    domain of the functional.
+                """
+                return self._op.derivative(x).adjoint(
+                    self._func.gradient(self._op(x)))
+
+        return CompositGradient()
+
+
+class FunctionalRightVectorMult(Functional, OperatorRightVectorMult):
+
+    """Expression type for the functional right vector multiplication.
+
+    Given a functional ``func`` and a vector ``y`` in the domain of ``func``,
+    this corresponds to the functional
+
+        ``(func * y)(x) = func(y*x)``.
+    """
+
+    def __init__(self, func, vector):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The domain of ``func`` must be a ``vector.space``.
+        vector : `LinearSpaceVector` in ``func.domain``
+            The vector to multiply by.
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        OperatorRightVectorMult.__init__(self, operator=func, vector=vector)
+
+        Functional.__init__(self, domain=func.domain)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self.vector * self.operator.gradient * self.vector
+
+    # TODO: can this be computed?
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+
+        Notes
+        -----
+        The nonsmooth solvers that make use of proximal operators in order to
+        solve a given optimization problem, see for example
+        `forward_backward_pd`, take a `proximal factory` as input. Note that
+        ``Functional.proximal`` is in fact a `proximal factory`.
+        """
+        raise NotImplementedError('there is no known expression for this')
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional.
```

Comment:
Commenting here again. Lots of copy-paste doc here that clutters the code. Since this inherits from another class where this doc already exists, rewriting it only adds complexity.